### ISSUE TITLE
gstreamer can't play HLS videos

### ISSUE THREAD

**Issue by andrein** [Homebrew/homebrew-core#162152]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
❯ brew config
HOMEBREW_VERSION: 4.2.7
ORIGIN: https://github.com/Homebrew/brew
HEAD: 6170b6297c61ef0b68814c708eeff2199d8b0d11
Last commit: 3 days ago
Core tap JSON: 08 Feb 20:12 UTC
Core cask tap JSON: 08 Feb 20:12 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.43.0 => /opt/homebrew/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.2.1-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: N/A
Rosetta 2: false

❯ brew doctor
Your system is ready to brew.
```

### Verification

- [X] My "`brew doctor` output" says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.
- [X] I searched for recent similar issues at https://github.com/Homebrew/homebrew-core/issues?q=is%3Aissue and found no duplicates.

### What were you trying to do (and why)?

I'm trying to play an HLS video using `gstreamer playbin3 uri=<...>`.

### What happened (include all command output)?

```
❯ gst-launch-1.0 playbin3 -v uri=https://test-streams.mux.dev/x36xhzz/x36xhzz.m3u8
objc[6512]: Class ResultReceiver is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.41/lib/libgtk-3.0.dylib (0x106af5e00) and /opt/homebrew/Cellar/gtk4/4.12.5/lib/libgtk-4.1.dylib (0x12dae32a8). One of the two will be used. Which one is undefined.
objc[6512]: Class GtkApplicationQuartzDelegate is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.41/lib/libgtk-3.0.dylib (0x106af5e28) and /opt/homebrew/Cellar/gtk4/4.12.5/lib/libgtk-4.1.dylib (0x12dae32d0). One of the two will be used. Which one is undefined.
objc[6512]: Class GNSMenuItem is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.41/lib/libgtk-3.0.dylib (0x106af5e78) and /opt/homebrew/Cellar/gtk4/4.12.5/lib/libgtk-4.1.dylib (0x12dae3320). One of the two will be used. Which one is undefined.
objc[6512]: Class GNSMenu is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.41/lib/libgtk-3.0.dylib (0x106af5ea0) and /opt/homebrew/Cellar/gtk4/4.12.5/lib/libgtk-4.1.dylib (0x12dae3348). One of the two will be used. Which one is undefined.
objc[6512]: Class FilterComboBox is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.41/lib/libgtk-3.0.dylib (0x106af5f18) and /opt/homebrew/Cellar/gtk4/4.12.5/lib/libgtk-4.1.dylib (0x12dae33c0). One of the two will be used. Which one is undefined.
objc[6512]: Class gdkCoreCursor is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.41/lib/libgdk-3.0.dylib (0x105e41628) and /opt/homebrew/Cellar/gtk4/4.12.5/lib/libgtk-4.1.dylib (0x12dae3410). One of the two will be used. Which one is undefined.
Setting pipeline to PAUSED ...
Pipeline is PREROLLING ...
/GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstDecodebin3:decodebin3-0: caps = video/x-raw(ANY); audio/x-raw(ANY); text/x-raw(ANY); subpicture/x-dvd; subpicture/x-dvb; subpicture/x-xsub; subpicture/x-pgs; closedcaption/x-cea-608; closedcaption/x-cea-708; application/x-onvif-metadata
/GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstURISourceBin:urisourcebin0: source = "\(GstCurlHttpSrc\)\ curlhttpsrc0"
/GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstURISourceBin:urisourcebin0/GstTypeFindElement:typefindelement0.GstPad:src: caps = application/x-hls
/GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstURISourceBin:urisourcebin0/GstHLSDemux:hlsdemux0.GstPad:sink: caps = application/x-hls
ERROR: from element /GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstURISourceBin:urisourcebin0: Your GStreamer installation is missing a plug-in.
Additional debug info:
../subprojects/gst-plugins-base/gst/playback/gsturisourcebin.c(762): new_demuxer_pad_added_cb (): /GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstURISourceBin:urisourcebin0:
Adaptive demuxer is not streams-aware, check your installation
ERROR: pipeline doesn't want to preroll.
Setting pipeline to NULL ...
Freeing pipeline ...
```

### What did you expect to happen?

Video should play.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew install gstreamer
```

**Comment by chenrui333** [comment]
I can see the same issue with 1.22.10 update

```
$ /opt/homebrew/Cellar/gstreamer/1.22.10/bin/gst-launch-1.0 playbin3 -v uri=https://test-streams.mux.dev/x36xhzz/x36xhzz.m3u8
objc[14062]: Class ResultReceiver is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.41/lib/libgtk-3.0.dylib (0x104c35e00) and /opt/homebrew/Cellar/gtk4/4.12.5/lib/libgtk-4.1.dylib (0x115ce32a8). One of the two will be used. Which one is undefined.
objc[14062]: Class GtkApplicationQuartzDelegate is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.41/lib/libgtk-3.0.dylib (0x104c35e28) and /opt/homebrew/Cellar/gtk4/4.12.5/lib/libgtk-4.1.dylib (0x115ce32d0). One of the two will be used. Which one is undefined.
objc[14062]: Class GNSMenuItem is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.41/lib/libgtk-3.0.dylib (0x104c35e78) and /opt/homebrew/Cellar/gtk4/4.12.5/lib/libgtk-4.1.dylib (0x115ce3320). One of the two will be used. Which one is undefined.
objc[14062]: Class GNSMenu is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.41/lib/libgtk-3.0.dylib (0x104c35ea0) and /opt/homebrew/Cellar/gtk4/4.12.5/lib/libgtk-4.1.dylib (0x115ce3348). One of the two will be used. Which one is undefined.
objc[14062]: Class FilterComboBox is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.41/lib/libgtk-3.0.dylib (0x104c35f18) and /opt/homebrew/Cellar/gtk4/4.12.5/lib/libgtk-4.1.dylib (0x115ce33c0). One of the two will be used. Which one is undefined.
objc[14062]: Class gdkCoreCursor is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.41/lib/libgdk-3.0.dylib (0x103f81628) and /opt/homebrew/Cellar/gtk4/4.12.5/lib/libgtk-4.1.dylib (0x115ce3410). One of the two will be used. Which one is undefined.
Setting pipeline to PAUSED ...
Pipeline is PREROLLING ...
/GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstDecodebin3:decodebin3-0: caps = video/x-raw(ANY); audio/x-raw(ANY); text/x-raw(ANY); subpicture/x-dvd; subpicture/x-dvb; subpicture/x-xsub; subpicture/x-pgs; closedcaption/x-cea-608; closedcaption/x-cea-708; application/x-onvif-metadata
/GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstURISourceBin:urisourcebin0: source = "\(GstCurlHttpSrc\)\ curlhttpsrc0"
/GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstURISourceBin:urisourcebin0/GstTypeFindElement:typefindelement0.GstPad:src: caps = application/x-hls
/GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstURISourceBin:urisourcebin0/GstHLSDemux:hlsdemux0.GstPad:sink: caps = application/x-hls
ERROR: from element /GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstURISourceBin:urisourcebin0: Your GStreamer installation is missing a plug-in.
Additional debug info:
../subprojects/gst-plugins-base/gst/playback/gsturisourcebin.c(762): new_demuxer_pad_added_cb (): /GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstURISourceBin:urisourcebin0:
Adaptive demuxer is not streams-aware, check your installation
ERROR: pipeline doesn't want to preroll.
Setting pipeline to NULL ...
Freeing pipeline ...
```

Looks like missing a plugin?

> ERROR: from element /GstPlayBin3:playbin3-0/GstURIDecodeBin3:uridecodebin3/GstURISourceBin:urisourcebin0: Your GStreamer installation is missing a plug-in.

**Comment by cho-m** [comment]
I guess this is an issue using the newer `hlsdemux2` included with GStreamer 1.22? I think we do build corresponding plugin which would be inside `adaptivedemux2`:
```console
❯ fd adaptivedemux2 /opt/homebrew/opt/gstreamer
/opt/homebrew/opt/gstreamer/lib/gstreamer-1.0/libgstadaptivedemux2.dylib
```

However, `gst-inspect-1.0` doesn't show the `dashdemux2`, `hlsdemux2`, or `mssdemux2` elements.
```console
❯ gst-inspect-1.0 adaptivedemux2
Plugin Details:
  Name                     adaptivedemux2
  Description              Adaptive Streaming 2 plugin
  Filename                 /opt/homebrew/Cellar/gstreamer/1.22.10/lib/gstreamer-1.0/libgstadaptivedemux2.dylib
  Version                  1.22.10
  License                  LGPL
  Source module            gst-plugins-good
  Documentation            https://gstreamer.freedesktop.org/documentation/adaptivedemux2/
  Source release date      2024-02-13
  Binary package           GStreamer Good Plug-ins source release
  Origin URL               https://github.com/Homebrew/homebrew-core


  0 features:

```

Comparing with legacy `hls` plugin, it looks like:
```console
❯ gst-inspect-1.0 hls
Plugin Details:
  Name                     hls
  Description              HTTP Live Streaming (HLS)
  Filename                 /opt/homebrew/Cellar/gstreamer/1.22.10/lib/gstreamer-1.0/libgsthls.dylib
  Version                  1.22.10
  License                  LGPL
  Source module            gst-plugins-bad
  Documentation            https://gstreamer.freedesktop.org/documentation/hls/
  Source release date      2024-02-13
  Binary package           GStreamer Bad Plug-ins
  Origin URL               https://github.com/Homebrew/homebrew-core

  hlsdemux: HLS Demuxer
  hlssink: HTTP Live Streaming sink
  hlssink2: HTTP Live Streaming sink

  3 features:
  +-- 3 elements

```

Not too sure what is happening yet. From quick glance, it doesn't seem like we are missing a dependency or manually disabling feature.

**Comment by cho-m** [comment]
This may be related to upstream issue https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/1171. Some extra discussion in upstream MR https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/2399

At least if I delete cache (e.g. `~/.cache/gstreamer-1.0`) and use a `DYLD_FALLBACK_LIBRARY_PATH` workaround to help find `libsoup`, the plugin loads correctly:
```console
❯ gst-inspect-1.0 adaptivedemux2
Plugin Details:
  Name                     adaptivedemux2
  Description              Adaptive Streaming 2 plugin
  Filename                 /opt/homebrew/Cellar/gstreamer/1.22.10/lib/gstreamer-1.0/libgstadaptivedemux2.dylib
  Version                  1.22.10
  License                  LGPL
  Source module            gst-plugins-good
  Documentation            https://gstreamer.freedesktop.org/documentation/adaptivedemux2/
  Source release date      2024-02-13
  Binary package           GStreamer Good Plug-ins source release
  Origin URL               https://github.com/Homebrew/homebrew-core

  dashdemux2: DASH Demuxer
  hlsdemux2: HLS Demuxer
  mssdemux2: Smooth Streaming demuxer (v2)

  3 features:
  +-- 3 elements
```

**Comment by cho-m** [comment]
Upstream has a fix in HEAD but commit doesn't apply cleanly. May need to wait for 1.24 release or a potential backport to 1.22.

**Comment by nirbheek** [comment]
@cho-m I've made a backport for you: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/6198

**Comment by cho-m** [comment]
> @cho-m I've made a backport for you: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/6198

Thanks. Sadly I didn't have a chance to work on applying diff.

We may just get the fix with 1.24.0 - #165176
